Add fleetctl debug connection command

mna opened the pull request:
Adds the `fleetctl debug connection` command to debug connection issues to the fleet server. It's a bit bigger than I would've liked, but it also includes some package organization refactoring that generated a bit of noise - the `secure/` and `orbit/pkg/certificate/` packages have moved to `/pkg/` at the top-level of the repo, as the place for packages shared between orbit and fleet.

Note that I might not be able to merge this until next week - if this is urgent and looks good, feel free to merge in my absence, otherwise I will address any comments (or merge if approved) at the start of next week.

* [x] Resolve hostname
* [x] Attempt tcp connection
* [x] Validate certificate if provided
* [x] Make API endpoint call and verify response
* [x] Use explicitly provided address if present, fallback to context's address

Fixes #1579

zwass commented on cmd/fleetctl/debug.go:
```diff
@@ -316,3 +322,144 @@ func debugArchiveCommand() *cli.Command {
+		Description: `Run a number of checks to debug a connection failure to the Fleet
+server.
+
+If <address> is provided, this is the address that is investigated,
+otherwise the address of the provided context is used, with
+the default context used if none is explicitly specified.`,
```
Nice, this seems useful to default to the server URL in the context.

zwass commented on cmd/fleetctl/debug.go:
```diff
@@ -316,3 +322,144 @@ func debugArchiveCommand() *cli.Command {
+				addr = c.Args().First()
+			}
+
+			// TODO: not correct: clientFromCLI requires a login token, which
+			// can only be obtained once the connection works. The context cannot apply
+			// for this command, use a provided address (as mentioned in the issue), and
+			// build the fleet client just before the API endpoint check.
+			_ = addr
```
Maybe `unauthenticatedClientFromCLI`?

mna replied:
Thanks, I missed that option. I ended up refactoring the clientFromCLI/unauthenticatedClientFromCLI a bit so I could get the context config separately (while reusing the existing validations in place regarding context existing and ensuring the config file exists), and override the address with the explicitly-provided one if present before creating the client (I still need the context config for the TLS insecure flag and root CA if configured).

zwass commented on cmd/fleetctl/debug.go:
```diff
@@ -316,3 +322,144 @@ func debugArchiveCommand() *cli.Command {
+	return nil
+}
+
+func checkFleetCert(ctx context.Context, timeout time.Duration, certPath, host string, ips []net.IP) error {
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	//	b, err := ioutil.ReadFile(cert)
+	//	if err != nil {
+	//		return errors.Wrap(err, "invalid certificate path")
+	//	}
+	//	// TODO: flag is the PEM, not the DER
+	//	xcert, err := x509.ParseCertificate(b)
+	//	if err != nil {
+	//		return errors.Wrap(err, "parsing x509 cert")
+	//	}
+	//	_ = xcert
+	return nil
+}
```
Check out https://github.com/fleetdm/fleet/blob/main/tools/osquery/fleet.crt for an example cert that we commonly use in development.

I also just remembered we have https://github.com/fleetdm/fleet/blob/main/orbit/pkg/certificate/certificate.go#L13-L64 which may be useful.

mna replied:
Thanks, that's super helpful! A couple things about the code in `certificate.go`:

* Is this ok for fleet (fleetctl command) to use an orbit package? I remember reading somewhere that those two could be used independently, maybe we want to keep that boundary in the code dependencies too? Maybe it's fine now that they live in the same repo though.
* I'm not super familiar with the various `tls.Config` fields, but reading the docs (https://pkg.go.dev/crypto/tls#Config) it seems like in `ValidateConnection` it should use `RootCAs` instead of `ClientCAs` - RootCAs is for the clients, to validate server certificates, while ClientCAs is for the servers, to validate client-provided certificates IF the ClientAuth policy is not NoClientCert (which is the default).

I can address this as part of the PR (if there's indeed anything to change).

zwass replied:
As discussed in chat, this package can be moved to a top-level `pkg` directory. Fleet and Orbit should be able to function independently, but it's fine for them to share code dependencies.

In regards to `ClientCAs`, looks like you are right. https://github.com/fleetdm/fleet/blob/main/orbit/pkg/certificate/certificate.go#L50 is probably the line that has the effect we were actually looking for. Please go ahead and make the appropriate fixes.

zwass reviewed (commented):
Looks like you are on the right track! I've tried to leave some comments where I think I can be helpful.

mna commented:
So, the output looks like this - happy to update/clarify if there are suggestions for improvements!

```
# no config file yet
$ fleetctl debug connection --config /tmp/nosuchfile
set the Fleet API address with: fleetctl config set --address https://localhost:8080
or provide an <address> argument to debug: fleetctl debug connection localhost:8080

# no config file yet, address provided (no server running there)
$ fleetctl debug connection --config /tmp/nosuchfile localhost:8081
Debugging connection to localhost; Configuration context: default; Root CA: (system); TLS: secure.
Success: can resolve host localhost.
Fail: dial server: dial tcp [::1]:8081: connect: connection refused

# no config file yet, address provided with server listening
$ fleetctl debug connection --config /tmp/nosuchfile localhost:8080
Debugging connection to localhost; Configuration context: default; Root CA: (system); TLS: secure.
Success: can resolve host localhost.
Success: can dial server at localhost:8080.
Fail: agent API endpoint: request failed: Post "https://localhost:8080/api/v1/osquery/enroll": x509: certificate signed by unknown authority

# existing config file with insecure set for localhost
$ fleetctl debug connection localhost:8080
Debugging connection to localhost; Configuration context: default; Root CA: (system); TLS: insecure.
Success: can resolve host localhost.
Success: can dial server at localhost:8080.
Success: agent API endpoints are available.

# unknown context
$ fleetctl debug connection -context martin localhost:8080
context "martin" is not found

# invalid hostname
$ fleetctl debug connection localhostz:8080
Debugging connection to localhostz; Configuration context: default; Root CA: (system); TLS: insecure.
Fail: resolve host: lookup localhostz: no such host

# with a valid certificate provided
$ fleetctl debug connection --fleet-certificate tools/osquery/fleet.crt localhost:8080
Debugging connection to localhost; Configuration context: default; Root CA: (system); TLS: insecure.
Success: can resolve host localhost.
Success: can dial server at localhost:8080.
Success: TLS certificate seems valid.
Success: agent API endpoints are available.

# with a valid certificate but not for this domain (i.e. localhost.localdomain == localhost 
# on my machine, but the certificate does not cover this domain name)
$ fleetctl debug connection --fleet-certificate tools/osquery/fleet.crt localhost.localdomain:8080
Debugging connection to localhost.localdomain; Configuration context: default; Root CA: (system); TLS: insecure.
Success: can resolve host localhost.localdomain.
Success: can dial server at localhost.localdomain:8080.
Fail: certificate: dial for validate: verify certificate: x509: certificate is valid for localhost, host.docker.internal, 127.0.0.1, not localhost.localdomain

# with an invalid certificate
$ fleetctl debug connection --fleet-certificate /tmp/cert.pemlocalhost:8080
Debugging connection to localhost; Configuration context: default; Root CA: (system); TLS: insecure.
Success: can resolve host localhost.
Success: can dial server at localhost:8080.
Fail: certificate: dial for validate: verify certificate: x509: certificate signed by unknown authority
```

chiiph commented on cmd/fleetctl/debug_test.go:
```diff
@@ -0,0 +1,182 @@
+			switch {
+			case res.code == 0:
+				panic(res.body)
+			case res.code < 0:
```
Sneaky :)

chiiph commented on server/service/client.go:
```diff
@@ -129,13 +131,24 @@ func (c *Client) doWithHeaders(verb, path, rawQuery string, params interface{},
+
 func (c *Client) Do(verb, path, rawQuery string, params interface{}) (*http.Response, error) {
+	return c.DoContext(context.Background(), verb, path,
+		rawQuery, params)
```
Nitpick: should go in 1 line only.

It would be nice to have a very opinionated `gofmt` for silly things like this to not happen.

zwass reviewed (changes requested):
This is going to be super helpful for our users debugging!

Code is looking good to me.

I notice a couple of potential issues in this test

```
$ ./build/fleetctl debug connection --fleet-certificate tools/osquery/fleet.crt localhost:8080
Debugging connection to localhost; Configuration context: secure; Root CA: (system); TLS: secure.
Success: can resolve host localhost.
Success: can dial server at localhost:8080.
Success: TLS certificate seems valid.
Fail: agent API endpoint: request failed: Post "https://localhost:8080/api/v1/osquery/enroll": x509: certificate signed by unknown authority
```

- [x] I think we need to use the cert chain in the "agent API endpoint" check
- [x] `Root CA: (system)` <- should this actually be something other than `system` since I provided a cert?

Also, I think we should ignore everything in the context if the address argument is provided. Otherwise it's hard to be sure which cert chain and insecure settings are being used.

mna commented:
>  I think we need to use the cert chain in the "agent API endpoint" check

Good point, I'll adjust the client when a cert is provided.

> Root CA: (system) <- should this actually be something other than system since I provided a cert?

I was still using the context to get the root CA and the skip insecure flag, and if the context's root is not specified, the command prints `(system)`.

> Also, I think we should ignore everything in the context if the address argument is provided. 

So if a cert is provided, I'd create an empty cert pool and add just that certificate, using this as root CA instead of whatever's in the context? And we'd never set the skip insecure flag if a certificate is provided?

mna commented:
I updated the PR according to @zwass ' comment and my response comment. Note that I updated the previous comment with the various command calls and the output it generates.

A couple of UX questions that still remain for me:

* should we raise an error if `--config` or `--context` is provided along with an explicit address? I think that'd be a yes, as we won't use any of those if an address is set.
* should we allow setting the `--fleet-certificate` flag when an address isn't provided? That would mean take the address from the config's context, but use this root cert instead? Seems a bit confusing, I think it'd be clearer if you either use it with a configured context, or use it with the explicit address and optional cert.

zwass commented:
> So that means if an address is provided but no cert, we just make an https request without setting any root ca on the http.Client (i.e. we use defaults, and we don't set the skip insecure flag), and if a cert is provided, we create an empty cert pool and add just that certificate, using this as root CA instead of whatever's in the context? And we'd never set the skip insecure flag in that case either (well I guess there's not much point to skip if we use that provided cert as root ca).

That sounds right.

> should we raise an error if --config or --context is provided along with an explicit address? 

Yes

> should we allow setting the --fleet-certificate flag when an address isn't provided? 

No I don't think we should allow that.

zwass reviewed (approved):
Thanks for working out the last issues!

mna commented:
:+1: I'll just clean up a bit some code that is now unused with the latest changes, and get this merged.
